### PR TITLE
Abandoned Crates (Ported from Baystation12)

### PR DESCRIPTION
**Pull request by pudl**
Adds abandoned crates to code. Originally coded for Goonstation by ISaidNo and included in his public release to Baystation, later reworked by @Kelenius.
- Play a game of mastermind to unlock a crate. If the crate isn't unlocked within 10 attempts (or emagged), a small explosion is made and the crate is destroyed.
- 50 "common" crates (with a 62.5% spawn chance), 36 "uncommon" crates (with a 25% spawn chance), 14 "rare" crates (with a 12.5% spawn chance).
- One abandoned crate added to the mining z-level.
- Across the space z-levels are several lootdroppers that have a 1/5 chance of creating an abandoned crate. 2 on z3, 2 and z4, 8 on z5.

Drop table here: http://pastebin.com/x7xEugPb

TL;DR:
- Best drops from common crates include a standard space suit with tape recorder and drill, or a bluespace beaker.
- Best drops from uncommon crates include 1-5 bluespace crystals, or a full ghetto surgery kit.
- Best drops from rare crates include a cursed or uncursed katana, 1 gatfruit seed, a hand tele, an all-access soulstone, or an alien brain.

Changelog coming.

### CODE REVIEW

**Comment by Chocobro** [comment]
>mastermind
great time to never be able to open these fucking things

**Comment by Chocobro** [comment]
>mastermind

great
time to never be able to open these fucking thing
:+1:

File: code/modules/mining/abandoned_crates.dm [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+			new /obj/item/weapon/storage/belt/champion(src)
+			new /obj/item/clothing/mask/luchador(src)
+		if(100)
+			new /obj/item/clothing/head/bearpelt(src)
+
+/obj/structure/closet/crate/secure/loot/togglelock(mob/user as mob)
+	if(locked)
+		user << "<span class='notice'>The crate is locked with a Deca-code lock.</span>"
+		var/input = input(usr, "Enter [codelen] digits.", "Deca-Code Lock", "") as text
+		if(in_range(src, user))
```
Comment (phil235) [inline code/modules/mining/abandoned_crates.dm]: also needs to check for user.incapacitated()

**Comment by pudl** [comment]
@Chocobro You have 10 tries so I think you should be able to get it.

File: code/modules/mining/abandoned_crates.dm (hunk 1 above)
Comment (pudl) [inline code/modules/mining/abandoned_crates.dm]: If you're going to do a line-by-line, tell me now, so I can fix everything in 1 big push.

Comment (phil235) [inline code/modules/mining/abandoned_crates.dm]: Yeah, I'm checking each line atm

**Comment by Alek2ander** [comment]
Maybe a chance of a crate mimic spawning instead?

**Comment by pudl** [comment]
@Alek2ander 
no

**Comment by pudl** [comment]
@phil235 So is everything good? Other than what you pointed out.

**Comment by phil235** [comment]
The code looks okay to me.

File: code/modules/mining/abandoned_crates.dm [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+		else
+			user << "<span class='notice'>You attempt to interact with the device using a hand gesture, but it appears this crate is from before the DECANECT came out.</span>"
+			return
+	else
+		return ..()
+
+/obj/structure/closet/crate/secure/loot/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	if(locked)
+		if (istype(W, /obj/item/weapon/card/emag))
+			user << "<span class='danger'>The crate's anti-tamper system activates!</span>"
```
Comment (Kelenius) [inline code/modules/mining/abandoned_crates.dm]: Why.

**Comment by PKPenguin321** [comment]
I think this is really cool, *but* there are some things in the rare crates that shouldn't be. Namely the gatfruit and alien brain. IIRC, some cybernetic implants are (or at least are going to be rebalanced to be) dependent on really high biotech levels, and the only way to get them that high are with alien brains. This is so that they're not normally obtainable and not a big factor in most rounds, but having them have a 12.5% chance of spawning and applying that chance 13 times every round opens them to be available almost every time somebody's willing to go out and explore space.

Gatfruits are infinite revolvers for the entire crew and generally shit on rounds and aren't fun, but this is honestly a more personal dislike (since they can already be obtained through xenobio). I guess it's up to you and the maintainers, but gatfruits really aren't something we should have more of.

Also, I would put the standard space suit in the uncommon crate and put the surgery kit in the common one, since I don't see the value that a ghetto surgery kit has over enabling space travel.

**Comment by RemieRichards** [comment]
@PKPenguin321 gatfruit doesn't (or didn't and honestly shouldn't) produce seeds.

**Comment by PKPenguin321** [comment]
Really?? It must have been changed at some points because the rounds I remember having gatfruit always had botany filled with hundreds of them. If that's true then I guess it's no problemo.

File: code/modules/mining/abandoned_crates.dm (hunk 2 above)
Comment (pudl) [inline code/modules/mining/abandoned_crates.dm]: Because being able to use your magic card to open up a crate that could have a katana in it is retarded.

**Comment by Iamgoofball** [comment]
here's the history on this:
I ported this from goon to tg, but the PR for that got denied
so i ported this from the denied pr to bay, they accepted it
now it's back here

so this is actually from goon, not bay

File: code/modules/mining/abandoned_crates.dm (hunk 2 above)
Comment (Iamgoofball) [inline code/modules/mining/abandoned_crates.dm]: no, the emag should be able to bypass the anti-tamper

revert this part

Comment (RemieRichards) [inline code/modules/mining/abandoned_crates.dm]: No, with the current push towards more lethality the emag should be included in that, it'll retain most of it's "win button" features, but on certain objects it'll fuck you up for trying to "cheat"

Comment (Iamgoofball) [inline code/modules/mining/abandoned_crates.dm]: what

i spent valuable TCs on this crap why should I get penalized for using it

**Comment by palpatine213** [comment]
Emag shouldn't just instantly fail the game for you.
Instead, it should make the next person to try and open the crate instantly activate the explosives. So a traitor can go around booby trapping the crates.

**Comment by Cheridan** [comment]
yeah gatfruits weren't really meant to actual usage, they're a joke/admin item that happens to be obtainable because xenobio is silly.

**Comment by Dorsisdwarf** [comment]
gatfruits absolutely make seeds, up to 4 packs from a single fruit

**Comment by Lumipharon** [comment]
remove the gatfruit chance. semi competent botanists will be rolling in hundreds of revolvers if they get that.

File: code/modules/mining/abandoned_crates.dm (hunk 2 above)
Comment (Anonus) [inline code/modules/mining/abandoned_crates.dm]: Mining actually has the least uses for the emag, so it should work and unlock the crate, or at least make the game absurdly simple(IE, one number needed)

**Comment by Anonus** [comment]
Replace the gatfruit with an actual gun, an actual gun is less dangerous than the gatfruit.

**Comment by AlexCrimson** [comment]
Considering the rarity of the Gatfruit, whats the harm? The chances of you getting that thing and actually being able to deliver it to Botany, then have them mass-produce the fruit are pretty low. It would be a rare silly thing to happen.

File: code/modules/mining/abandoned_crates.dm (hunk 2 above)
Comment (pudl) [inline code/modules/mining/abandoned_crates.dm]: I like the booby trapping idea someone had. The reason I made the emag not work is because Goon changed the emag to exploding the crates, and player feedback was generally positive (as people were sick of traitors getting good items with no effort from crates.)

**Comment by pudl** [comment]
@PKPenguin321 I want to keep the alien brain. I'm going to switch the space suit and surgery kit.
@Cheridan @Anonus @Lumipharon  I'm changing the gatfruit to a stechkin with 1 magazine and a balaclava.
@Iamgoofball In-code I specifically wrote the ISaidNo coded this originally. However, considering this port uses the "mastermind" minigame from Bay, I would consider it a Baystation12 port.

**Comment by Tokiko1** [comment]
Isn't it far easier to get gatfruits from silver cores currently than it would be by opening crates? I've seen competant xenobiologists getting multiple fruits in a single round and yet nobody seems to complain about it being unbalanced.

Then again, a gun would probably still be more useful to someone opening a crate and expecting an instant reward. A stechkin seems pretty weak though.

**Comment by Lumipharon** [comment]
stechkins are great though. tiny size, 30 brute per shot.

**Comment by pudl** [comment]
For whatever reason, clicking a crate doesn't call the game's code, and it just fucking opens without any trouble. It's pissing me off. Debug message shows that it doesn't even get past "if(locked)". Which, in that case, should call the message saying that you wave your hand on it, but not even that is called. The entire string of code is ignored.

File: code/modules/mining/abandoned_crates.dm [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+			new /obj/item/weapon/storage/belt/champion(src)
+			new /obj/item/clothing/mask/luchador(src)
+		if(100)
+			new /obj/item/clothing/head/bearpelt(src)
+
+/obj/structure/closet/crate/secure/loot/togglelock(mob/user as mob)
+	if(locked)
+		user << "<span class='notice'>The crate is locked with a Deca-code lock.</span>"
+		var/input = input(usr, "Enter [codelen] digits.", "Deca-Code Lock", "") as text
+		if(in_range(src, user) || !user.incapacitated())
```
Comment (phil235) [inline code/modules/mining/abandoned_crates.dm]: the user incapacitated check needs to be separate, otherwise the message about using hand gesture will show up .

**Comment by phil235** [comment]
The crate opens because /obj/structure/closet/crate/secure/attack_hand doesn't call togglelock. See https://github.com/tgstation/-tg-station/blob/909fcdc9a532f452817f8da85d4b67f46fe156dd/code/game/objects/structures/crates_lockers/crates.dm#L224

File: code/modules/mining/abandoned_crates.dm (hunk 3 above)
Comment (pudl) [inline code/modules/mining/abandoned_crates.dm]: You can't interact with the crate through a border window. Just checked.

Comment (Kelenius) [inline code/modules/mining/abandoned_crates.dm]: Check if you can open the code window, move it behind a window, and enter the code through window.

Comment (phil235) [inline code/modules/mining/abandoned_crates.dm]: But can you open the input window then go behind a border window and then enter a code?

File: code/modules/mining/abandoned_crates.dm [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+			new /obj/item/weapon/storage/belt/champion(src)
+			new /obj/item/clothing/mask/luchador(src)
+		if(100)
+			new /obj/item/clothing/head/bearpelt(src)
+
+/obj/structure/closet/crate/secure/loot/attack_hand(mob/user as mob)
+	if(locked)
+		user << "<span class='notice'>The crate is locked with a Deca-code lock.</span>"
+		var/input = input(usr, "Enter [codelen] digits.", "Deca-Code Lock", "") as text
+		if(in_range(src, user) && !user.incapacitated())
```
Comment (Miauw62) [inline code/modules/mining/abandoned_crates.dm]: ```user.canUseTopic(src)```

Comment (Miauw62) [inline code/modules/mining/abandoned_crates.dm]: actually make that user.canUseTopic(src, 1)

Comment (pudl) [inline code/modules/mining/abandoned_crates.dm]: Will do, mind explaining to me the difference?

Comment (phil235) [inline code/modules/mining/abandoned_crates.dm]: one difference is that mob without dexterity can't use it (aliens, monkeys I think)

Comment (pudl) [inline code/modules/mining/abandoned_crates.dm]: I removed the incapacitated because I assumed Miauw wanted me to do the whole line, if not then I'll put it back.

**Comment by Miauw62** [comment]
the extra 1 is called "be_close" in the arguments list, I assume that speaks for itself.

**Comment by freerealestate** [comment]
I like this idea quite a lot, but as seems to be the usual practice with /tg/, it seems too difficult to get ahold of anything fun to bother with. 

You have to become spaceworthy, travel through space, find one of the 12 spawn locations, hope you got the 1/5 chance for the crate to spawn, otherwise travel to a new location and again hope you won the 20% chance, then win a game of mastermind (or get blown up) and then hope you ended up with a decent crate and not two pairs of shorts or any of the other innumerable common or junk items in the droplist.

IMO the crates themselves should either be easier to find (e.g. chance to spawn in maint), or the droplist made better to incentivise finding them. Because after the novelty wears off, I doubt many people would continue to look for them.

**Comment by Lumipharon** [comment]
Not everything has to have such consistanly fantastic loot that everyone goes out of their way every round to get them.

**Comment by freerealestate** [comment]
On average, only two of these crates will be hidden in space per round.

Looking at the list of best items, the only items that seem worth the trip are the rare ones. The best items in the common and uncommon crates can be found more easily elsewhere, aren't that fun or useful, or only benefit certain jobs.

So there's only a 1/8 chance of getting something useful or fun that you can't find more easily elsewhere. 

The reward doesn't seem to compensate the effort needed to interact with them. Just my opinion, though.

**Comment by Iamgoofball** [comment]
why don't you make these pop up on mining?

gives miners something fun to play with

**Comment by pudl** [comment]
They are already on mining.

**Comment by Miauw62** [comment]
space being a magic place full of godly loot that people will want every round is, considering all the other factors, a pretty bad thing to do right now, because somebody in space is not really contributing to the round in any way.
im ok with this, though.

**Comment by pudl** [comment]
@Miauw62 To be honest, everyone besides a very, very small select few want expansion of space content, and want space exploration to be an actually attractive thing. I don't want to try to talk you into agreeing with that, since the PR isn't for that, but I just wanted to throw that out there.
If, somehow, coderbus ever comes to perfect harmony and decides that space content is good, then crates can be utilized to help that.

**Comment by Iamgoofball** [comment]
@pudl nono

i mean they should show up from mining

as a random drop from mining away at crap ala goon and bay's implementation

that's more fun

**Comment by pudl** [comment]
@Iamgoofball That's the dumbest thing I've ever heard. I can understand it on goon since you're pulling in massive blocks of scrap and rock bits, but here we have a single solid asteroid. Crates just being lodged in rock makes no sense.

**Comment by Iamgoofball** [comment]
@pudl it was like that when they had the preset asteroids too

also making sense really doesn't matter here

**Comment by palpatine213** [comment]
Mining already has randomly generated loot rooms. Why does it need even more loot?

**Comment by Iamgoofball** [comment]
@palpatine213 because 1. the loot rooms are crap
2. this is fun

**Comment by pudl** [comment]
Can this be merged? Thanks.

**Comment by phil235** [comment]
changelog please.

**Comment by pudl** [comment]
@phil235 Done.
